tests/0113-cooperative_rebalance.cpp

```diff
@@ -26,311 +26,1714 @@
+        Test::poll_once(c2);
+      }
+
+      Test::Say("Consumers are subscribed to one partition each\n");
+      done = true;
+    }
+  }
+
+  Test::Say("Closing consumer 1\n");
+  c1->close();
+  Test::Say("Closing consumer 2\n");
+  c2->close();
+
+  /* There should be no assign rebalance_cb calls on close */
+  if (rebalance_cb1.assign_call_cnt != 2)
+    Test::Fail(tostr() << "Expecting consumer 1's assign_call_cnt to be 2 not: " << rebalance_cb1.assign_call_cnt);
+  if (rebalance_cb2.assign_call_cnt != 2)
+    Test::Fail(tostr() << "Expecting consumer 2's assign_call_cnt to be 2 not: " << rebalance_cb2.assign_call_cnt);
+
+  if (rebalance_cb1.revoke_call_cnt != 2)
+    Test::Fail(tostr() << "Expecting consumer 1's revoke_call_cnt to be 2 not: " << rebalance_cb1.assign_call_cnt);
+  if (rebalance_cb2.revoke_call_cnt != 2)
+    Test::Fail(tostr() << "Expecting consumer 2's revoke_call_cnt to be 2 not: " << rebalance_cb2.assign_call_cnt);
+
+  if (rebalance_cb1.lost_call_cnt != 1)
+    Test::Fail(tostr() << "Expecting consumer 1's lost_call_cnt to be 1, not: " << rebalance_cb1.lost_call_cnt);
+  if (rebalance_cb2.lost_call_cnt != 1)
+    Test::Fail(tostr() << "Expecting consumer 2's lost_call_cnt to be 1, not: " << rebalance_cb2.lost_call_cnt);
+
+  delete c1;
+  delete c2;
+}
+
+
+
+/* Check behavior when:
+ *   1. Consumer (librdkafka) subscribes to two topics (2 and 6 partitions).
+ *   2. Consumer (java) subscribes to the same two topics.
+ *   3. Consumer (librdkafka) unsubscribes from the two partition topic.
+ *   4. Consumer (java) process closes upon detecting the above unsubscribe.
+ *   5. Consumer (librdkafka) will now be subscribed to 6 partitions.
+ *   6. Close librdkafka consumer.
+ */
+
+static void o_java_interop() {
+  Test::Say("Executing o_java_interop\n");
+
+  std::string topic_name_1 = Test::mk_topic_name("0113-cooperative_rebalance", 1);
+  std::string topic_name_2 = Test::mk_topic_name("0113-cooperative_rebalance", 1);
+  std::string group_name = Test::mk_unique_group_name("0113-cooperative_rebalance");
+  test_create_topic(NULL, topic_name_1.c_str(), 2, 1);
+  test_create_topic(NULL, topic_name_2.c_str(), 6, 1);
+
+  DefaultRebalanceCb rebalance_cb;
+  RdKafka::KafkaConsumer *c = make_consumer("C_1", group_name, false, "cooperative-sticky", &rebalance_cb);
+  test_wait_topic_exists(c->c_ptr(), topic_name_1.c_str(), 10*1000);
+  test_wait_topic_exists(c->c_ptr(), topic_name_2.c_str(), 10*1000);
+
+  Test::subscribe(c, topic_name_1, topic_name_2);
+
+  bool done = false;
+  bool java_started = false;
+  bool changed_subscription = false;
+  bool changed_subscription_done = false;
+  int java_pid;
+  while (!done) {
+    Test::poll_once(c);
+
+    if (Test::partition_count(c) == 8 && !java_started) {
+      Test::Say("librdkafka consumer assigned to 8 partitions\n");
+      string bootstrapServers = get_bootstrap_servers();
+      const char **argv = (const char **)rd_alloca(sizeof(*argv) * (1 + 1 + 1 + 1 + 1 + 1));
```

Comment:
Since the allocation is of fixed size, why not just use fixed size array and use an initializer list?

Reply:
changed